installinator/src/dispatch.rs

```diff
@@ -208,33 +209,140 @@ impl InstallOpts {
+                                installinator_doc.artifact.clone(),
+                                installinator_doc_id.hash,
+                            )
+                            .await?;
+
+                            // Read the document as JSON.
+                            //
+                            // Going via the reader is slightly less efficient
+                            // than operating purely in-memory, but serde_json
+                            // doesn't have a good way to pass in a BufList
+                            // directly.
+                            let json: InstallinatorDocument =
+                                serde_json::from_reader(buf_list::Cursor::new(
+                                    &installinator_doc.artifact,
+                                ))
+                                .context(
+                                    "error deserializing \
+                                     installinator document",
+                                )?;
+
+                            // Every valid installinator document must have the
+                            // host phase 2 and control plane hashes.
+                            let mut host_phase_2_hash = None;
+                            let mut control_plane_hash = None;
+                            for artifact in &json.artifacts {
+                                match artifact.kind {
+                                    InstallinatorArtifactKind::HostPhase2 => {
+                                        host_phase_2_hash = Some(artifact.hash);
+                                    }
+                                    InstallinatorArtifactKind::ControlPlane => {
+                                        control_plane_hash =
+                                            Some(artifact.hash);
+                                    }
+                                }
+                            }
+
+                            // Return an error if either the host phase 2 or the
+                            // control plane artifacts are missing.
+                            let mut missing = Vec::new();
+                            if host_phase_2_hash.is_none() {
+                                missing.push(
+                                    InstallinatorArtifactKind::HostPhase2,
+                                );
+                            }
+                            if control_plane_hash.is_none() {
+                                missing.push(
+                                    InstallinatorArtifactKind::ControlPlane,
+                                );
+                            }
+                            if !missing.is_empty() {
+                                bail!(
+                                    "installinator document missing \
+                                     required artifacts: {:?}",
+                                    missing
+                                );
+                            }
```

Comment:
Is there any advantage to trying to return an error over `bail!`?

Reply:
Ah this is within the context of a single update-engine step, so it's not an immediate system exit -- installinator will report this up to wicket before exiting.